### PR TITLE
OBM-3057 Add pickaxe id to manifest request

### PR DESCRIPTION
**Pull request by cja769**
Adding the pickaxe id to the manifest request allows the backend to override the global version of pickaxe per pickaxe

---

> [!NOTE]
> **Medium Risk**
> Changes how and how often manifest is fetched and what the backend sees for version selection; behavior is well-tested but affects production upgrades and couples to a large `java-api` version jump.
> 
> **Overview**
> **Upgrade polling** now fetches `/api/manifest` **once per scale identifier** instead of a single shared request. For each identifier, `UpgradeTask` looks under that identifier’s dist for a `foreman-pickaxe-*` install, reads `pickaxeId` from `conf/pickaxe.yml`, and appends it as a URL-encoded `pickaxe_key` query parameter when present (using `&` if the base URL already has a query string). If there is no pickaxe, no id, or the YAML is invalid/empty, the manifest URL is left unchanged so the backend can still serve the global manifest.
> 
> **Dependencies:** `foreman.version` in `pom.xml` is bumped from `1.38.0` to `1.86.0`.
> 
> **Tests:** Large new JUnit coverage for `UpgradeTask` (install/upgrade/skip paths and `pickaxe_key` behavior), plus tests for `Upgrader`, `WatchDog`, `AppImpl`, `FileUtils`, `VersionFactoryImpl`, `AppManifest`, and `AppFolder`.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 6913d87bf618e506ed908aecc1cb31ca4d12976b. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -12,7 +12,7 @@
 
     <groupId>mn.foreman</groupId>
     <artifactId>windows-agent</artifactId>
-    <version>1.5.1</version>
+    <version>1.6.0</version>
```
Comment (delawr0190) [inline pom.xml]: [Breaking] The agent version is being bumped from `1.5.1` to `1.6.0` — both non-SNAPSHOT. Develop is expected to stay on a `*-SNAPSHOT` version between releases (e.g. `1.6.0-SNAPSHOT` if you're opening the 1.6.0 cycle). Merging new code at a released coordinate is visible to any consumer that pulls the artifact by exact version, and it short-circuits the usual release/snapshot dance. The bump on develop should be to `1.6.0-SNAPSHOT`; the release-to-`1.6.0` cut belongs in the release PR that targets `main`.

File: src/main/java/mn/foreman/windowsagent/upgrade/UpgradeTask.java [hunk 2]
```diff
@@ -88,39 +100,134 @@ public UpgradeTask(
...
+                    final String dist =
+                            this.agentDist + File.separator + identifier;
+
+                    final String pickaxeKey =
+                            findPickaxeIdInDist(dist).orElse("");
+                    final String url =
+                            buildManifestUrl(this.appsManifest, pickaxeKey);
+
+                    final AppManifest[] appManifests =
+                            this.restTemplate.getForObject(
```
Comment (delawr0190) [inline src/main/java/mn/foreman/windowsagent/upgrade/UpgradeTask.java:119]: [Edge case] Previously the manifest endpoint was GET'd once per `check()` and the result was applied across all identifiers. After this change it's GET'd once per identifier — so a tenant running e.g. 10 scales now generates 10× the manifest-endpoint load per check cycle. Two things worth confirming: (1) is the endpoint cached/cheap enough on the backend that this fan-out is safe at scale, and (2) when the same `pickaxe_key` appears for multiple identifiers (common when one user runs many scales), are we OK re-fetching the same response N times rather than caching per-key within a single `check()`?

File: src/main/java/mn/foreman/windowsagent/upgrade/UpgradeTask.java [hunk 3]
```diff
@@ -88,39 +100,134 @@ public UpgradeTask(
...
+     * @return The URL with the pickaxe key embedded as a query parameter,
+     *         or the base URL unchanged when no key is available.
+     */
+    private static String buildManifestUrl(
+            final String base,
+            final String pickaxeKey) {
+        if (pickaxeKey.isEmpty()) {
+            return base;
+        }
+        final String separator = base.contains("?") ? "&" : "?";
```
Comment (delawr0190) [inline src/main/java/mn/foreman/windowsagent/upgrade/UpgradeTask.java:161]: [Edge case] The `base.contains("?")` test handles `&` vs `?` for the separator, but it doesn't guard against the base URL already carrying a `pickaxe_key=` query param (e.g. a hardcoded override in someone's config, or a legacy install). In that case the request goes out with two `pickaxe_key=` values; which one wins is backend-dependent (Spring `@RequestParam String` takes the first, some other frameworks take the last). If `appsManifest` is always a bare URL in production today, this is theoretical — but worth either stripping any existing `pickaxe_key` before appending, or asserting the invariant in config validation.

File: src/main/java/mn/foreman/windowsagent/upgrade/UpgradeTask.java [hunk 4]
```diff
@@ -88,39 +100,134 @@ public UpgradeTask(
...
+     *
+     * @return The pickaxe ID, if one could be resolved.
+     */
+    private static Optional<String> findPickaxeIdInDist(final String dist) {
+        final File distRoot = new File(dist);
+        if (!distRoot.isDirectory()) {
+            return Optional.empty();
+        }
+        final File[] appDirs =
+                distRoot.listFiles(file ->
```
Comment (delawr0190) [inline src/main/java/mn/foreman/windowsagent/upgrade/UpgradeTask.java:184]: [Edge case] `File.listFiles(...)` returns entries in an order the JVM/filesystem does not guarantee. During an in-progress upgrade, both `foreman-pickaxe-<old>` and `foreman-pickaxe-<new>` will live side-by-side under the identifier dist until the old version is cleaned up; the loop here returns the first one whose `conf/pickaxe.yml` yields a non-empty `pickaxeId`. If the invariant is that every sibling `foreman-pickaxe-*` shares the same `pickaxeId` (install-scoped, not version-scoped), this is fine and worth a one-line comment to make that explicit. If the `pickaxeId` could ever diverge between sibling versions (e.g. after a re-pair that rewrites the yml on only the newer install), the manifest fetched here becomes non-deterministic. Picking the highest-version folder explicitly would remove the ambiguity.

**Review by delawr0190 — DISMISSED** [review]
Per-pickaxe manifest fetches: derives a `pickaxe_key` from each identifier's installed `pickaxe.yml` (`foreman-pickaxe-*/conf/pickaxe.yml`) and appends it as a query param to the manifest endpoint URL. Also bumps the agent version from `1.5.1` to `1.6.0` and `foreman.version` from `1.38.0` to `1.86.0`.

### Develop-branch state
- develop `pom.xml` is currently on `1.5.1` (released, non-SNAPSHOT). Per the convention that develop must remain on a `*-SNAPSHOT` version between releases, a snapshot bump must land on develop before any code-edit PR can merge into it — otherwise merging new code onto a released coordinate is visible to any consumer that pulls the artifact by exact version.

### Inline findings
- [Breaking] pom.xml:15 — proposed agent version is also non-SNAPSHOT
- [Edge case] src/main/java/mn/foreman/windowsagent/upgrade/UpgradeTask.java:119 — manifest fetch moves from 1× per check to N× per check
- [Edge case] src/main/java/mn/foreman/windowsagent/upgrade/UpgradeTask.java:161 — separator selection if base URL already carries `pickaxe_key`
- [Edge case] src/main/java/mn/foreman/windowsagent/upgrade/UpgradeTask.java:184 — undefined directory listing order when multiple `foreman-pickaxe-*` folders coexist

I'll re-review on push.

File: src/main/java/mn/foreman/windowsagent/upgrade/UpgradeTask.java (hunk 2 above)
Comment (cja769) [reply src/main/java/mn/foreman/windowsagent/upgrade/UpgradeTask.java:119]: 1. I'm not sure if anyone is even using the scale and colo features of this since it's no longer necessary
2. Even if it is being used, it's so infrequent that I don't think it matters.
3. I'm pretty sure windows agent allows you to run multiple pickaxes which would have multiple keys

Long story short, this way is far more convenient and the downsides are beyond negligible

File: src/main/java/mn/foreman/windowsagent/upgrade/UpgradeTask.java (hunk 3 above)
Comment (cja769) [reply src/main/java/mn/foreman/windowsagent/upgrade/UpgradeTask.java:161]: This is a nonsensical theoretical case

File: src/main/java/mn/foreman/windowsagent/upgrade/UpgradeTask.java (hunk 4 above)
Comment (cja769) [reply src/main/java/mn/foreman/windowsagent/upgrade/UpgradeTask.java:184]: There's not a case where a new version would have a different pickaxe version in the yaml. This is so obviously case, idk why it's asking for documentation

File: src/main/java/mn/foreman/windowsagent/upgrade/UpgradeTask.java [hunk 5]
```diff
@@ -88,39 +100,134 @@ public UpgradeTask(
...
+                                    .collect(Collectors.toList());
+
+                    checkManifest(
+                            identifierString,
+                            windowsApps,
+                            dist,
+                            versions.computeIfAbsent(
+                                    identifierString,
+                                    s -> new ConcurrentHashMap<>()));
+                });
```
[the remark this answers is not in the document]
Comment (cja769) [reply src/main/java/mn/foreman/windowsagent/upgrade/UpgradeTask.java:139]: Yeah I'd argue this is an improvement

File: pom.xml [hunk 6]
```diff
@@ -12,7 +12,7 @@
 
     <groupId>mn.foreman</groupId>
     <artifactId>windows-agent</artifactId>
-    <version>1.5.1</version>
+    <version>1.6.0-SNAPSHOT</version>
```
Comment (delawr0190) [inline pom.xml:15]: [Acknowledgement] Previously flagged the proposed agent version as non-SNAPSHOT (1.6.0); the bump to `1.6.0-SNAPSHOT` here resolves that at the PR-branch level — confirmed in this diff hunk. The remaining base-branch state concern (develop is still on `1.5.1`) is captured in the review body's Develop-branch state section, not on this line.

File: src/main/java/mn/foreman/windowsagent/upgrade/UpgradeTask.java (hunk 2 above)
Comment (delawr0190) [inline src/main/java/mn/foreman/windowsagent/upgrade/UpgradeTask.java:119]: [Acknowledgement] Previously flagged the per-identifier manifest fetch as a potential N× fan-out concern relative to the old single-fetch design. Your reply that windows-agent installs typically run a small number of identifiers, and that the previous single-fetch couldn't carry a per-pickaxe key at all, resolves the practical-scale half of the question — confirmed for OBM's expected per-agent identifier counts. (The separate exception-propagation half on this same lambda — one `getForObject` failure short-circuiting the rest of the `forEach` — is independently raised by the other reviewer on this PR, so I'm not duplicating it here.)

File: src/main/java/mn/foreman/windowsagent/upgrade/UpgradeTask.java (hunk 3 above)
Comment (delawr0190) [inline src/main/java/mn/foreman/windowsagent/upgrade/UpgradeTask.java:161]: [Acknowledgement] Previously flagged the separator-only `?`/`&` check as not guarding against an existing `pickaxe_key` already present in the base URL. Your reply that production `appsManifest` is always a bare `/api/manifest` URL (no pre-existing query params) resolves the practical concern — confirmed against the configuration shape we ship today. If a future configuration path ever lets operators paste a pre-built URL, this is the place to revisit.

File: src/main/java/mn/foreman/windowsagent/upgrade/UpgradeTask.java (hunk 4 above)
Comment (delawr0190) [inline src/main/java/mn/foreman/windowsagent/upgrade/UpgradeTask.java:184]: [Acknowledgement] Previously flagged the non-deterministic `listFiles` order with sibling `foreman-pickaxe-*` folders. Your point that the `pickaxeId` in `pickaxe.yml` is install-scoped (so all sibling-version yml files carry the identical value) resolves the determinism question — under that invariant the pick-first-non-empty-yml strategy here is consistent, and the listing order doesn't matter.

**Review by delawr0190 — DISMISSED** [review]
Adds per-pickaxe manifest targeting: derives a `pickaxe_key` from each identifier's installed `foreman-pickaxe-*/conf/pickaxe.yml` (parsed with SnakeYAML `SafeConstructor`) and appends it as a URL-encoded `pickaxe_key` query param on the `/api/manifest` fetch. Also bumps the agent to `1.6.0-SNAPSHOT` and `foreman.version` from `1.38.0` to `1.86.0`, plus extensive unit coverage for the new pickaxe-key URL logic, `WatchDog`, `AppImpl`, `VersionFactoryImpl`, `FileUtils`, and `AppManifest`.

### Develop-branch state
- develop `pom.xml` is currently on `1.5.1` (released, non-SNAPSHOT). Per the convention that develop must remain on a `*-SNAPSHOT` version between releases, a snapshot bump must land on develop before any code-edit PR can merge into it — otherwise merging new code onto a released coordinate is visible to any consumer that pulls the artifact by exact version. This PR carries a `1.6.0-SNAPSHOT` bump in its own diff, but the gate measures develop's current state at review time, not the proposed post-merge state — a develop-only bump PR (or splitting this one) should land first.

### Inline findings
- [Acknowledgement] pom.xml:15 — prior 1.6.0 non-SNAPSHOT concern resolved by this PR's bump to 1.6.0-SNAPSHOT
- [Acknowledgement] src/main/java/mn/foreman/windowsagent/upgrade/UpgradeTask.java:119 — accepted author's response on manifest fetch fan-out
- [Acknowledgement] src/main/java/mn/foreman/windowsagent/upgrade/UpgradeTask.java:161 — accepted author's response on base URL shape
- [Acknowledgement] src/main/java/mn/foreman/windowsagent/upgrade/UpgradeTask.java:184 — accepted author's install-scoped `pickaxeId` claim

I'll re-review on push.

File: src/main/java/mn/foreman/windowsagent/upgrade/UpgradeTask.java (hunk 2 above)
Comment (delawr0190) [inline src/main/java/mn/foreman/windowsagent/upgrade/UpgradeTask.java:119]: [Design] This moves the manifest fetch from one shared request per `check()` cycle to one request per identifier. `Upgrader.check()` builds this task over `0..scale-1` for private installs and over every colo sub-client id (Upgrader.java:108-150), so on a colo host a single `/api/v2/manifests` poll becomes one request per sub-client every `${upgrade.check.frequency}` cycle — the manifest backend's request rate now scales with sub-client count instead of being constant per agent. This is inherent to keying the manifest per pickaxe with the current endpoint; flagging it so the backend's capacity for the multiplied rate is a deliberate choice rather than an accidental side effect (a batched endpoint accepting all keys in one request would avoid the fan-out).

**Review by delawr0190 — APPROVED** [review]
Fetches the app manifest once per identifier and appends that identifier's installed pickaxe id (read from `foreman-pickaxe-*/conf/pickaxe.yml`) as a `pickaxe_key` query param so the backend can override the pickaxe version per install; also bumps `foreman.version` 1.38.0→1.86.0 and adds unit coverage across several previously-untested classes.

### Inline findings
- `[Design] src/main/java/mn/foreman/windowsagent/upgrade/UpgradeTask.java:119` — manifest fetch is now per-identifier, so backend request rate scales with colo sub-client count.

### Verified
- `readPickaxeId` parses with `new Yaml(new SafeConstructor())` (UpgradeTask.java:217), which only produces standard YAML types and rejects arbitrary-class tags — not the default `Constructor`, so untrusted `pickaxe.yml` content cannot instantiate arbitrary Java objects.
- The `pickaxe_key` param is additive: an empty/absent key returns the base URL unchanged (`buildManifestUrl`, UpgradeTask.java:158-159), and `pickaxe_key` is an existing platform parameter (used in foreman-apps `foreman-pickaxe` WebSocketService), so an unrecognized key falls back to the global manifest rather than breaking the fetch.
- The `foreman.version` 1.38.0→1.86.0 bump compiles under green CI; its only consumers (`ForemanApi`/`JdkWebUtil` in Configuration.java and Upgrader.java) are unchanged and are not on the manifest-fetch path.
- Per-identifier URL isolation is asserted by `usesPerIdentifierPickaxeKey`, and `urlEncodesPickaxeKey` confirms the key is URL-encoded so a raw `&` cannot inject extra query params.
- The null-manifest-body path skips only the affected identifier (UpgradeTask.java:122-125), and the no-pickaxe / empty-id / malformed-yaml paths are each tested and tolerated.

No breaking concerns identified after independent review.

**Review by delawr0190 — APPROVED** [review]
